vault/logical_system.go

```diff
@@ -2955,6 +2956,37 @@ func (b *SystemBackend) handleMetrics(ctx context.Context, req *logical.Request,
 	return b.Core.metricsHelper.ResponseForFormat(format), nil
 }
 
+func (b *SystemBackend) handleInFlightRequestData(_ context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	now := time.Now()
+	currentInFlightReqMap := make(map[string]interface{})
+	syncMapRangeResult := true
+	b.Core.inFlightReqMap.Range(func(key, value interface{}) bool {
+		v, ok := value.(*InFlightReqData)
+		if !ok {
+			syncMapRangeResult = false
+			return false
+		}
+		// don't report the request to the in-flight-req path
+		if v.ReqPath != "/v1/sys/in-flight-req" {
+			v.Duration = fmt.Sprintf("%v microseconds", now.Sub(v.StartTime).Microseconds())
```

Comment:
I'm still on the fence about whether we want to have a `duration` field, or if the starttime is good enough.  But assuming we keep duration, I would rather we not modify the struct pointed to by the sync.Map value.  For one thing, if we have two simultaneous requests to this endpoint, we could have a data race as both attempt to populate this field.  But also just on principle, it feels wrong to set that field except in this handler's response: in the sync.Map the request is still in flight, so it doesn't have a duration yet.

Reply:
I see. I am wondering why a data race could happen considering we are using sync.Map. I think there should be some internal locking in place. Though, I do see I should not have changed the struct pointed to by the sync.Map, will fix. 
We could change the "duration" name to something more meaningful though, like "timeSinceStarted" or even the time at which we are taking the snapshot of the in-flight requests like "snapshot_time" ( I am bad with choosing field/variable names). Something like snapshot_time would be a convenient as a user won't have to consider the communication overhead time. This will come in handy for debugging specially when a lot of hops exist between the client and the server.

Reply:
sync.Map will protect you from races involving access to the map, but once you've got the value for an entry, any concurrent access to the record it is pointing to must be protected.